### ISSUE TITLE
(Metabase 0.50.5 is unavailable) +CrashOnOutOfMemoryError cpu problem

### ISSUE THREAD

**Issue by alperengunes** [metabase/metabase#44359]
### Describe the bug

![image](https://github.com/metabase/metabase/assets/38505390/0ddbd2a9-e0d2-4dee-b7a1-24d9fc766287)

### Information about your Metabase installation

```JSON
Docker

POSTGRESQL
METABASE
```

444002525bab   metabase/metabase:latest   "/app/run_metabase.sh"   2 hours ago     Up 27 minutes (healthy)   0.0.0.0:8080->3000/tcp, :::8080->3000/tcp   metabase
2b09134d6196   postgres:latest            "docker-entrypoint.s…"   10 months ago   Up 27 minutes             5432/tcp                                    postgres

### Severity

10

**Comment by uladzimirdev** [comment]
@alperengunes could you please add more details about your configuration?

**Comment by paoliniluis** [comment]
@alperengunes please post the logs, specs... at least you're not running Metabase correctly there since you're not probably tweaking the heap limit

**Comment by alperengunes** [comment]
In 49.x updates I don't have this problem, but in 50.x updates I have this problem. There are no different changes I have made.

**Comment by paoliniluis** [comment]
@alperengunes if you don't send us what we need, we won't be able to identify the problem

**Comment by alperengunes** [comment]
> @alperengunes if you don't send us what we need, we won't be able to identify the problem

[metabase-tail-10000-logs.txt](https://github.com/user-attachments/files/15888226/metabase-tail-10000-logs.txt)

**Comment by alperengunes** [comment]
not only this but I have been getting expression error since I updated my reports.

![image](https://github.com/metabase/metabase/assets/38505390/d9c79323-0668-4678-b8b5-bc484c6ad069)
![image](https://github.com/metabase/metabase/assets/38505390/33fe077b-010e-47ba-b4c4-f720c70ee898)

**Comment by rob-pomelo** [comment]
also seeing significant performance regression since going from 0.49 -> 0.50, cpu usage spiking, will try get some logs

**Comment by LukeAbell** [comment]
Also seeing performance regression and CPU spikes.

**Comment by perivamsi** [comment]
@LukeAbell any diagnostics you can share with us that can help debug this?

**Comment by LukeAbell** [comment]
What can I share? It seems like it's mostly affecting large dashboards with a lot of questions. Looks like it's the metadata endpoint mostly.

![Screenshot 2024-06-18 at 02 05 30 PM@2x](https://github.com/metabase/metabase/assets/3343530/89d9c67e-9a7d-4a50-ab8a-c213b930b02a)
![Screenshot 2024-06-18 at 02 07 23 PM@2x](https://github.com/metabase/metabase/assets/3343530/cc272018-8ac1-4b3b-b18c-757b9d7ab54f)

**Comment by rob-pomelo** [comment]
also running in gke and have had similar cpu spike on upversion just like ^

**Comment by uladzimirdev** [comment]
@rob-pomelo https://www.metabase.com/docs/latest/troubleshooting-guide/diagnostic-info

**Comment by LukeAbell** [comment]
@perivamsi Just emailed support logs for the dashboard.

**Comment by paoliniluis** [comment]
@LukeAbell @rob-pomelo and @alperengunes can you add MB_SQL_PARSING_ENABLED=false, reboot and then let us know if this keeps happening?

@alperengunes 
2024-06-18 14:33:00,039 WARN malli.fn :: Invalid output
ERROR middleware.add-source-metadata :: Error determining expected columns for query
clojure.lang.ExceptionInfo: Error preprocessing query in metabase.query_processor.preprocess
Caused by: clojure.lang.ExceptionInfo: Invalid query

please post the questions that spit those lines

**Comment by matotias** [comment]
Not sure if it's the same issue, but we started seeing huge spikes in the database, this updates seem to be the issue 
<img width="1095" alt="image" src="https://github.com/metabase/metabase/assets/29378537/95a6c509-0651-428a-8ff0-d2feaa8366a4">

**Comment by paoliniluis** [comment]
@matotias what version did you upgrade from?

**Comment by paoliniluis** [comment]
@matotias and what engine/sizing are you using as well

**Comment by LukeAbell** [comment]
@paoliniluis Didn't seem to change much. I think it's permission or sandbox related because it's much faster as a super admin vs a normal user that's sandboxed using full app embed.

**Comment by LukeAbell** [comment]
Check out the difference in time between sandboxed (right) and super admin (left).
![Screenshot 2024-06-18 at 03 26 53 PM@2x](https://github.com/metabase/metabase/assets/3343530/8c005d26-37ee-40d2-8447-f0d934730157)

**Comment by matotias** [comment]
> @matotias and what engine/sizing are you using as well

We went from `0.49.12` to `0.50.4` and the spikes started, then we updated again to `0.50.5` but we're still seeing them

We use postgres, 4 cpu 16 gigs of ram

**Comment by paoliniluis** [comment]
@LukeAbell I'm trying to reproduce what you're seeing and I'm wondering how it's taking 4 seconds to get a 30kb response. Can you describe a little more the queries? how are queries built (based on questions or models?). Are tables big? many joins?

**Comment by paoliniluis** [comment]
> > @matotias and what engine/sizing are you using as well
> 
> We went from `0.49.12` to `0.50.4` and the spikes started, then we updated again to `0.50.5` but we're still seeing them
> 
> We use postgres, 4 cpu 16 gigs of ram

4 cpu and 16 gigs of ram is the postgres sizing right?

**Comment by LukeAbell** [comment]
@paoliniluis want to hop on a quick video chat?

**Comment by paoliniluis** [comment]
@LukeAbell give me a couple of hours to try to go to the call with some answers, can you make it tomorrow first time in the morning?

**Comment by LukeAbell** [comment]
@paoliniluis Tomorrow at 10am EST work?

**Comment by LukeAbell** [comment]
> @LukeAbell I'm trying to reproduce what you're seeing and I'm wondering how it's taking 4 seconds to get a 30kb response. Can you describe a little more the queries? how are queries built (based on questions or models?). Are tables big? many joins?

Queries are based on questions (no models involved). Not a ton of joins. Using bigquery- table has about 800k rows.

Only 6 tables are returned in that query_metadata call.

**Comment by paoliniluis** [comment]
Thanks @LukeAbell, do you have observability in the app db to see if there's something weird there?

**Comment by LukeAbell** [comment]
Yep- doesn't seem like anything weird there. Less than 20% utilization. Running MySQL.

**Comment by matotias** [comment]
> > > @matotias and what engine/sizing are you using as well
> > 
> > 
> > We went from `0.49.12` to `0.50.4` and the spikes started, then we updated again to `0.50.5` but we're still seeing them
> > We use postgres, 4 cpu 16 gigs of ram
> 
> 4 cpu and 16 gigs of ram is the postgres sizing right?

Yes!

**Comment by alperengunes** [comment]
> @alperengunes could you click on the "download diagnostics info" and send over that file to us after reviewing that there is no sensitive information there?

checking

**Comment by alperengunes** [comment]
@perivamsi
[expression_logs.txt](https://github.com/user-attachments/files/15892255/expression_logs.txt)

**Comment by matotias** [comment]
Update on our case: we went back to v0.49.12 and the database load is normal again

**Comment by alperengunes** [comment]
> @LukeAbell @rob-pomelo and @alperengunes can you add MB_SQL_PARSING_ENABLED=false, reboot and then let us know if this keeps happening?
> 
> @alperengunes 2024-06-18 14:33:00,039 WARN malli.fn :: Invalid output ERROR middleware.add-source-metadata :: Error determining expected columns for query clojure.lang.ExceptionInfo: Error preprocessing query in metabase.query_processor.preprocess Caused by: clojure.lang.ExceptionInfo: Invalid query
> 
> please post the questions that spit those lines

not working :'(

**Comment by paoliniluis** [comment]
@matotias is your app db a postgres or mysql?

**Comment by matotias** [comment]
postgres

**Comment by paoliniluis** [comment]
ok we have some strings to pull from now

**Comment by naveenthontepu** [comment]
Metabase is crashing every 3 to 4 hours in our case also. There is no significant jump in the CPU utilisation but the RAM utilisation has significantly increased and is consuming the complete 4 gb ram we have on the metabase server. 

Looking at the tasks running on our server by running "htop" we are seeing that there are too many tasks metabase is keeping active even when there is nobody using metabase.

**Comment by perivamsi** [comment]
@naveenthontepu could you send us the output of `htop` and logs of the tasks that Metabase is keeping active?

**Comment by paoliniluis** [comment]
Also: any specific environment variable that you’re using?

**Comment by naveenthontepu** [comment]
![Screenshot 2024-06-19 at 5 29 24 PM](https://github.com/metabase/metabase/assets/13089476/eaa034d3-ab07-4338-8fa7-9d6604d3990e)

There are no environment variables that we are using. 

the command we are using to run metabase specifically mentions to use only 2gb of the 4gb ram but it is not being considered and using it 4gb and crashing. In the above screenshot also we can see that metabase is consuming 71% memory which is ~3gb.

**Comment by noahmoss** [comment]
@LukeAbell what version of MySQL are you on?

**Comment by LukeAbell** [comment]
@noahmoss 8.0.31

**Comment by psneha716** [comment]
Hi, 
I tried disabling query parsing using MB_SQL_PARSING_ENABLED=false but it's of no help. Metabase is still crashing with OOM.
Version: v0.50.1

**Comment by paoliniluis** [comment]
We’re working on this as we speak

**Comment by perivamsi** [comment]
@LukeAbell we have some changes coming that that might improve the performance. would you be able to upgrade to 50.7 and test? which version are you on right now? is it easy for you to downgrade if 50.7 is still slow for you?

**Comment by LukeAbell** [comment]
@perivamsi yep just let me know when there's a tagged release. Or do I need to test on a custom branch?

**Comment by perivamsi** [comment]
no need to test a custom branch, I'll give you a proper release link once ready

**Comment by shrey-locad** [comment]
exactly same problem since the upgrade to 0.50. memory and cpu spiking intermittently bringing down the whole container every 3-4 hours. We will give it 1-2 more days before downgrading back to 0.49

**Comment by perivamsi** [comment]
@shrey-locad sorry about that! are you open to having a call that can help us debug this better? as I said, we are working on this but it would be helpful to have additional data points around the memory and CPU spikes. also, hi from another IITB alum!

**Comment by cheyuriy** [comment]
Similar issue with CPU spikes after upgrading from 0.49.x to 0.50.6. 
Using Postgres 14 as a DB. Metabase is deployed on 2CPU-12GB in GCE and DB is in CloudSQL.

Upgrade was in the end of 06/24.
 
<img width="1102" alt="Screenshot 2024-06-25 at 11 12 36" src="https://github.com/metabase/metabase/assets/6412917/7254ce97-41d1-451a-8e07-fd2effa4c6b0">

**Comment by shrey-locad** [comment]
@perivamsi - haha, that's a pleasant surprise. Hi 👋🏽 

It might be hard to debug over a call, especially since the crashes are unpredictable and intermittent. We have been unable to identify a specific trigger either. The latest crash was in the middle of the night where we're pretty sure none of our users were online. How about we gave you ssh access to the instance instead? Call works too. Either way. shrey.jain@golocad.com

**Comment by perivamsi** [comment]
That works, we just need to look at the logs. I'll email you.

**Comment by perivamsi** [comment]
@LukeAbell @shrey-locad @cheyuriy can you all please upgrade to 50.7 which has some fixes for the performance issues?

https://github.com/metabase/metabase/releases/tag/v0.50.7

**Comment by LukeAbell** [comment]
@perivamsi I upgraded but had to quickly downgrade because it completely broke the application. Downgrading back to v1.50.6 fixed the issue.

Assets wouldn't load giving an error:

```
Refused to execute script from 'HOST/app/dist/runtime.ba18610064ff9fba6745.js' because its MIME type ('') is not executable, and strict MIME type checking is enabled.
```

**Comment by perivamsi** [comment]
We're on it

**Comment by bshepherdson** [comment]
That sounds like https://github.com/metabase/metabase/issues/39449. If force-refresh doesn't work, try another browser?

It does seem like it was a transitory issue in the past.

**Comment by LukeAbell** [comment]
@perivamsi  @bshepherdson Completely cleared cache and that worked. Unfortunately not much of an improvement in performance (if any) 😔

- Query metadata taking 4-6 seconds sandboxed 
- Queries taking 3-5 seconds sandboxed
- CPU usage still spiking like crazy loading a single dashboard

**Comment by uladzimirdev** [comment]
@LukeAbell just for the record, what kind of cache have you cleared? what CDN do you use?

**Comment by LukeAbell** [comment]
@uladzimirdev I had to clear my local browser cache to resolve the issue. We have metabase behind a google cloud CDN but that didn't be the issue.

**Comment by psneha716** [comment]
Hi folks,
Could someone who has already downgraded their metabase version, help me out with the downgrading process? Tried downgrading, but it isn't working. Any help would be greatly appreciated. Thanks!

**Cluster configuration:**

ECS cluster with 2 tasks.
Task CPU: 3800 units (3.711 vCPU)
Task memory: 3700 MiB (3.613 GB)


**Steps that I followed for downgrading** (followed instructions from [here](https://www.metabase.com/docs/latest/installation-and-operation/upgrading-metabase#rolling-back-an-upgrade)):

- Stopped the container running v0.50.1 metabase instance

- Ran the following on the EC2 instance directly:

```
$docker run --rm metabase/metabase:v0.50.1 "migrate down"
Warning: environ value jdk-11.0.23+9 for key :java-version has been overwritten with 11.0.23
2024-06-12 21:08:25,427 INFO metabase.util :: Maximum memory available to JVM: 1.9 GB
2024-06-12 21:08:27,762 INFO util.encryption :: Saved credentials encryption is DISABLED for this Metabase instance. 🔓
 For more information, see https://metabase.com/docs/latest/operations-guide/encrypting-database-details-at-rest.html
2024-06-12 21:08:28,566 WARN db.env :: WARNING: Using Metabase with an H2 application database is not recommended for production deployments. For production deployments, we highly recommend using Postgres, MySQL, or MariaDB instead. If you decide to continue to use H2, please be sure to back up the database file regularly. For more information, see https://metabase.com/docs/latest/operations-guide/migrating-from-h2.html
2024-06-12 21:08:32,573 INFO driver.impl :: Registered abstract driver :sql  🚚
2024-06-12 21:08:32,579 INFO driver.impl :: Registered abstract driver :sql-jdbc (parents: [:sql]) 🚚
2024-06-12 21:08:32,585 INFO metabase.util :: Load driver :sql-jdbc took 32.7 ms
2024-06-12 21:08:32,586 INFO driver.impl :: Registered driver :h2 (parents: [:sql-jdbc]) 🚚
2024-06-12 21:08:32,717 INFO driver.impl :: Registered driver :mysql (parents: [:sql-jdbc]) 🚚
2024-06-12 21:08:32,744 INFO driver.impl :: Registered driver :postgres (parents: [:sql-jdbc]) 🚚
2024-06-12 21:08:34,401 INFO metabase.core ::
Metabase v0.50.1 (cc4ca82)

Copyright © 2024 Metabase, Inc.

Metabase Enterprise Edition extensions are NOT PRESENT.
2024-06-12 21:08:34,657 INFO db.setup :: Setting up Liquibase...
2024-06-12 21:08:35,245 INFO db.liquibase :: Updating liquibase table to reflect consolidated changeset filenames
2024-06-12 21:08:35,254 INFO db.liquibase :: No migration lock found.
2024-06-12 21:08:35,255 INFO db.liquibase :: Migration lock acquired.
2024-06-12 21:08:35,259 INFO db.setup :: Liquibase is ready.
2024-06-12 21:08:35,262 INFO db.liquibase :: No migration lock found.
2024-06-12 21:08:35,263 INFO db.liquibase :: Migration lock acquired.
2024-06-12 21:08:35,264 INFO db.liquibase :: Rolling back app database schema to version 49
```

- Tried spinning up a container using v0.49.15, but getting the following error:

```
2024-06-12 21:10:56,380 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...
--
2024-06-12 21:10:56,382 INFO db.setup :: �[36mVerifying postgres Database Connection ...�[0m
2024-06-12 21:10:56,724 INFO db.setup :: Successfully verified PostgreSQL 13.13 application database connection. ✅
2024-06-12 21:10:56,725 INFO db.setup :: �[36mChecking if a database downgrade is required...�[0m
2024-06-12 21:10:56,834 ERROR middleware.log :: �[31mGET /api/health 503 287.2 µs (0 DB calls)
{:status "initializing", :progress 0.3}
�[0m
2024-06-12 21:10:56,844 ERROR middleware.log :: �[31mGET /api/health 503 182.0 µs (0 DB calls)
{:status "initializing", :progress 0.3}
�[0m
2024-06-12 21:10:57,344 ERROR metabase.core :: Metabase Initialization FAILED
clojure.lang.ExceptionInfo: �[31mERROR: Downgrade detected.�[0m
Your metabase instance appears to have been downgraded without a corresponding database downgrade.
You must run `java -jar metabase.jar migrate down` from version 50.
Once your database has been downgraded, try running the application again.
```

**Comment by noahmoss** [comment]
Hi @psneha716 — I don't see any errors in the downgrading logs that indicate specifically why it didn't work. But I did fix an issue related to downgrading in Metabase 50.7, the latest release. Could you try starting that version up, and then running the downgrade command with it?

**Comment by psneha716** [comment]
Hi @noahmoss ,
It did not work. Created a new issue [here](https://github.com/metabase/metabase/issues/44784).

**Comment by psneha716** [comment]
Seeing frequent OOM crashes in v0.50.7 too. Don't see any improvement.

**Comment by cheyuriy** [comment]
@bshepherdson it's a slow climb till the limit 
graphs for virtual and resident memory usage for today with 2 restarts
<img width="1122" alt="Screenshot 2024-06-27 at 10 48 24" src="https://github.com/metabase/metabase/assets/6412917/7738b706-ea54-4599-866a-541e661f9a9c">
<img width="1108" alt="Screenshot 2024-06-27 at 10 49 13" src="https://github.com/metabase/metabase/assets/6412917/821bfcdb-6f8c-4d6d-968b-3a1094979353">

Also I can see constantly increasing number of "total active threads" in logs. Both time it climbed from 0 to ~16500 before fail.
`2024-06-27 06:14:26,632 DEBUG middleware.log :: [32mPOST /api/dashboard/242/dashcard/8869/card/9623/query 202 [ASYNC: completed] 4.1 s (22 DB calls) App DB connections: 1/100 Jetty threads: 3/100 (15 idle, 0 queued) (16,637 total active threads) Queries in flight: 7 (0 queued) {:metabase-user-id 910}[0m`

**Comment by bshepherdson** [comment]
Ah, thanks for those! Good catch with the runaway thread count, that's a strong signal.

**Comment by bshepherdson** [comment]
Just to keep these dots connected: there's a fix which only applies to users of MySQL (for querying, not as Metabase's appdb) that **greatly** reduces thread pool pressure while querying MySQL. It was fixed in #44719 which merged a short while ago and will be in 50.8.

If everyone with this OOM issue is querying MySQL, then I think that's highly likely to be the issue. If some here are seeing OOMs without talking to MySQL, then there's more to find. I'm still investigating in any case, but please chime in about MySQL!

**Comment by cheyuriy** [comment]
@bshepherdson we don't query MySQL. Only postgres and BigQuery

**Comment by bshepherdson** [comment]
@cheyuriy okay, thanks. I'll keep digging for what the runaway threads might be coming from.

**Comment by bshepherdson** [comment]
Anyone who is having this issue who is able to send CLI commands to your instance, can you run `jstack -l PID` and send the output? That should give a dump of all the threads in the running process with names and stacks, which might be huge with 16k but should illuminate which source of threads is leaking, how they got created, and what's holding them open.

(@cheyuriy since I know you specifically have runaway thread spawns)

**Comment by cheyuriy** [comment]
@bshepherdson sry, we've restarted it minutes ago. I'll run this command later, when it will be close to this number again

**Comment by bshepherdson** [comment]
Never mind! We managed to find one of our cloud instances on 50.x with a runaway thread pool, and get that thread dump from it. This is an issue with dangling timeout futures when querying Bigquery. This should be fixed tomorrow (Friday).

**Comment by psneha716** [comment]
Hey Guys,
Even after downgrading to v0.49.12, I'm seeing a lot of crashes due to OOM. Is there any stable version that I can go back to?

**Comment by bshepherdson** [comment]
Unless there was a sharp spike to 100% there that was missed by the metrics, why would it crash at 44%? The right side of that graph looks like a major GC to me.

Are you still seeing runaway thread counts? I don't know when the (BigQuery-specific) thread leak that's causing these OOMs was introduced, but given how many people are seeing it on 50.x it's hard to believe it's also been in 49.12 for months but it hasn't been reported until now.

If it is happening in 49.x we can backport the fix (#44860) to the 49 branch as well. It should land in the 50 branch in 30m or less and we'll get a 50.8 release out today. (Fixing this OOM issue plus some dashboard load time slowness, especially for sandboxing users.)

**Comment by psneha716** [comment]
@bshepherdson How do I diagnose the issue?

**Comment by psneha716** [comment]
I increased the heap space of the application to 70% of the total RAM, it hasn't crashed since then but the memory utilisation is just consistently going up. It's just a matter of time before it crashes.

<img width="1318" alt="image" src="https://github.com/metabase/metabase/assets/152379423/920566e0-14d8-4eef-a4e6-10b65c68b0cf">

**Comment by perivamsi** [comment]
Everyone who reported OOM crashes (especially if you are using BigQuery), we released a fix for this issue with v50.8. Please upgrade: https://github.com/metabase/metabase/releases/tag/v0.50.8

**Comment by alperengunes** [comment]
> Everyone who reported OOM crashes (especially if you are using BigQuery), we released a fix for this issue with v50.8. Please upgrade: https://github.com/metabase/metabase/releases/tag/v0.50.8

Thank you @perivamsi

**Comment by shrey-locad** [comment]
upgraded to 0.50.8. Will keep you posted

**Comment by MathiasGr** [comment]
Hi there, we upgraded to 0.50.8 yesterday and just experienced the CPU craziness this morning again.

The CPU usage seems to exponentially increase until everything is stalling.

<img width="507" alt="image" src="https://github.com/metabase/metabase/assets/18528990/ed1b8a34-3c7d-427f-8cb6-f9e0ae8e2b0e">

**Comment by cheyuriy** [comment]
Memory issues are less severe with 0.50.8, however they are still present. And yes, we can observe CPU spikes too.

**Comment by perivamsi** [comment]
@cheyuriy can you share more details about the memory issues? Can you do a thread dump and send it to us? `jstack -l <pid>`

**Comment by perivamsi** [comment]
Understood. Any insights you can provide into the resource usage?

**Comment by perivamsi** [comment]
@MathiasGr are you able to provide a thread dump? anything that helps us see what is going on during the CPU spikes will be helpful.

**Comment by perivamsi** [comment]
@shrey-locad how is your resource usage looking after the upgrade?

**Comment by perivamsi** [comment]
@MathiasGr the CPU graph you showed is that of the Metabase process or the BigQuery database?

**Comment by cheyuriy** [comment]
Looks like it's possible to get thread dump by sending `kill -3` to the java process inside the container. I'll try to do it later, when it will spawn a lot of threads

**Comment by MathiasGr** [comment]
> @MathiasGr ~the CPU graph you showed is that of the Metabase process or the BigQuery database?~
> 
> I understand It is Metabase

Indeed, it's Metabase

**Comment by MathiasGr** [comment]
I can try but these events happen randomly and when they do the entire pod stalls so it's almost impossible to get anything from it.

**Comment by cheyuriy** [comment]
[dump.txt](https://github.com/user-attachments/files/16090472/dump.txt)
Thread dump after ~9 hours uptime

**Comment by bshepherdson** [comment]
@cheyuriy thanks! Quick analysis: 8000+ threads in the `clojure-agent-send-off-pool-NNN` group, all blocking inside the `execute-bigquery` logic. That tracks with our theory that the fix in 50.8 #44860 only got some of the cases. Dan's #45105 seems to address the remaining cases in his load testing.

That doesn't explain the CPU utilization spikes, though, at least not directly.

Some questions for anyone seeing the CPU spikes:
- Do they occur on some rhythm, or at random times?
- If a rhythm, does it track with some other event?
    - Especially with Metabase's sync process, which is scheduled in your admin settings.
- Does memory usage correlate with the spikes at all? (If so, the spikes could really be a burst of GC thrashing, ending with an OOM crash and recovery after a restart.)

**Comment by bshepherdson** [comment]
@MathiasGr See my questions [above](https://github.com/metabase/metabase/issues/44359#issuecomment-2207269022). You say "randomly", but the graph shows they happen pretty consistently at the same time every day, which makes me suspect they're linked with some event such as syncing, or certain requests to Metabase. Of course that timing may just be a coincidence, too.

**Comment by MathiasGr** [comment]
> @MathiasGr See my questions [above](https://github.com/metabase/metabase/issues/44359#issuecomment-2207269022). You say "randomly", but the graph shows they happen pretty consistently around the same time every day, which makes me suspect they're linked with some event such as syncing, or certain requests to Metabase. Of course that timing may just be a coincidence, too.

Sorry Im currently on holiday but will take a look next week

**Comment by cheyuriy** [comment]
> Some questions for anyone seeing the CPU spikes:
> 
> Do they occur on some rhythm, or at random times?
> If a rhythm, does it track with some other event?
> Especially with Metabase's sync process, which is scheduled in your admin settings.
> Does memory usage correlate with the spikes at all? (If so, the spikes could really be a burst of GC thrashing, ending with an OOM crash and recovery after a restart.)

Can't see any pattern here. It happens when it has already consumed a lot of memory, but not correlated with anything else. It's even not the peak hour of Metabase usage in our company.

**Comment by calherries** [comment]
@MathiasGr @cheyuriy You could help us narrow down the issue by disabling fingerprinting and field value scanning for all databases, and seeing if the issue persists. These have been the source of similar issues in the past. You can do this by going to Admin > Databases and setting one of these options:
![image](https://github.com/metabase/metabase/assets/39073188/e1c25d7e-1f66-49b7-a53f-ea88bb3d98c1)

**Comment by calherries** [comment]
Thanks, that helps. @bshepherdson that implies it's not fingerprinting or field value scanning that's causing their issue.

**Comment by bshepherdson** [comment]
#45253 will fix the thread leak that leads to OOMs. I hope that it will also fix the unresponsive Metabase issues, because it appears the CPU spike and unresponsiveness may be due to GC thrashing as memory is nearly exhausted.

I'm doing a bit more profiling of general CPU usage with BigQuery before and after #45253; it removes some CPU/allocation overhead while processing individual queries, but I'm not sure how significant the impact is. Last week I wasn't able to reproduce any significant CPU usage while eg. loading a dashboard full of BQ queries, but I'm going to try a couple more approaches now.

**Comment by perivamsi** [comment]
With https://github.com/metabase/metabase/pull/45536, I would like to close this issue and open any new specific issues as needed.

**Comment by paoliniluis** [comment]
@MathiasGr can you test the new release?

**Comment by paoliniluis** [comment]
same @alperengunes @cheyuriy @matotias @shrey-locad @psneha716 @rob-pomelo, please all upgrade to 50.13 and post your results

**Comment by shrey-locad** [comment]
upgraded to 50.13. No issues since last 16 hours.

**Comment by paoliniluis** [comment]
We found that the queries in 50 are taking longer, and probably there's also more memory pressure than on previous versions. We'll start working on those 2 as soon as we can

**Comment by perivamsi** [comment]
@rob-pomelo could you please email me at vamsi@metabase.com? would like to get more details from you so we can fix your issue. ideally if you have time to meet and screen share, that would help a lot.

**Comment by ranquild** [comment]
Closed by https://github.com/metabase/metabase/pull/45881

**Comment by dmeremyanin** [comment]
Just my 2 cents after our Metabase upgrade today.

We updated from 0.49.21 to 0.50.15 and observed significantly higher CPU usage along with a decrease in performance. Metabase is deployed in Kubernetes using a configuration of 2 pods with 12 CPU and 8GB memory limit. Is started hitting the CPU limit, becoming unresponsive and triggering restarts due to the liveness probe.

We rolled back to version 0.49.21, and it started working perfectly fine again.

<img width="1303" alt="Screenshot 2024-07-24 at 1 46 04 PM" src="https://github.com/user-attachments/assets/a1302efb-a829-4467-9d0e-e50955fa28d6">

**Comment by paoliniluis** [comment]
@dmeremyanin can you contact us to our support email? We would like you test specific versions of the product if that’s fine with you. We would like to profile your deployment to find the causes of this

**Comment by perivamsi** [comment]
Should be fixed by https://github.com/metabase/metabase/pull/46108

**Comment by perivamsi** [comment]
@dmeremyanin we will release v50.16 with a fix for the performance issue. please test against that version and let us know if if you are still facing the CPU spikes.

**Comment by dmeremyanin** [comment]
Apologies for the delayed response and thank you for the prompt reaction. We will definitely test v50.16 and get back here with feedback. Thanks

**Comment by perivamsi** [comment]
@dmeremyanin 50.16 is out. Could you please test?

**Comment by MikitaKurlovich** [comment]
Hi there! Still have the issue on v50.16

docker compose:
```
  metabase-postgres:
    image: postgres:16.3
    container_name: metabase-postgres
    environment:
      POSTGRES_DB: metabase
      POSTGRES_USER: <metabase_user>
      POSTGRES_PASSWORD: <metabase_password>
    volumes:
      - metabase-postgres-data:/var/lib/postgresql/data
    ports:
      - "5432:5432”

  metabase:
    restart: unless-stopped
    image: metabase/metabase:v0.50.16
    ports:
      - 3001:3000
    environment:
      MB_DB_TYPE: postgres
      MB_DB_DBNAME: metabase
      MB_DB_PORT: 5432
      MB_DB_USER: <metabase_user>
      MB_DB_PASS: <metabase_password>
      MB_DB_HOST: metabase-postgres
    volumes:
      - ./metabase-data:/metabase-data
    depends_on:
      - metabase-postgres
```      
Diagnostic Info:
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.24+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.24",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.24+8",
    "os.name": "Linux",
    "os.version": "4.15.0-213-generic",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "16.3 (Debian 16.3-1.pgdg120+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    },
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-07-25",
      "tag": "v0.50.16",
      "hash": "28de9df"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```
      
<img width="1470" alt="Screenshot 2024-07-25 at 16 50 23" src="https://github.com/user-attachments/assets/a073585b-ae14-4cec-bcce-0aa48fbf6644">

**Comment by paoliniluis** [comment]
hi @MikitaKurlovich your screenshot doesn't show cpu or memory pressure, can you please specify your problem?

**Comment by dmeremyanin** [comment]
@perivamsi we've updated to v50.16. There's no traffic on it at the moment, but we are monitoring the load.

**Comment by perivamsi** [comment]
Thanks @dmeremyanin, keep me posted!

**Comment by paoliniluis** [comment]
@dmeremyanin any news?

**Comment by dmeremyanin** [comment]
Everything looks good so far, but there is still not much traffic.

**Comment by dmeremyanin** [comment]
Overall, the same issues: very high CPU utilization, which leads to freezing. We've rolled back to v0.49.21. Is there anything I can share that can help debug the problem?

<img width="1301" alt="Screenshot 2024-07-26 at 8 02 53 AM" src="https://github.com/user-attachments/assets/ecf4899e-f035-4032-9c23-734ac4501c9e">

**Comment by perivamsi** [comment]
@dmeremyanin mind emailing me at vamsi@metabase.com? would like to meet and debug if possible.

**Comment by dmeremyanin** [comment]
@perivamsi sure, I just sent you an email.

**Comment by V-Legendre** [comment]
Hey guys, I am observing the same problems with version v1.50.17.
Had to rollback to v49 to make Metabase available again.

![Screenshot 2024-07-31 at 11 02 44](https://github.com/user-attachments/assets/d95d049b-50c4-48d7-93f6-5f37549727f4)


Also, one SQL query executed by MB migrations was hanging indefinitely (+2hours) during the upgrade. I had to kill by hand and MB started nonetheless.

**Comment by perivamsi** [comment]
@V-Legendre are you open to sharing more diagnostic info over email or slack to help us fix the CPU spike issue? Please email me at vamsi@metabase.com if you are.

**Comment by apostoltego** [comment]
Hey just to add to this thread as we're seeing similar complaints. One thing that we observed was that admin users don't experience slowness. As an experiment I worked with a user that wasn't an admin and we were seeing CPU spikes on all dashboards and after making them an admin the issue went away. 
![Screenshot 2024-08-02 at 13 17 16](https://github.com/user-attachments/assets/b552f7cb-0f80-4106-a870-7c9f4a57e2c9)


Diagnostics ingo:
```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.24+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.24",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.24+8",
    "os.name": "Linux",
    "os.version": "5.10.219-208.866.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "clickhouse"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "13.10 (Ubuntu 13.10-1.pgdg22.04+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    },
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-07-26",
      "tag": "v0.50.17",
      "hash": "afd6b17"
    },
    "settings": {
      "report-timezone": "UTC"
    }
  }
}
```

**Comment by perivamsi** [comment]
@apostoltego are the non-admin users sandboxed or not?

**Comment by apostoltego** [comment]
@perivamsi apologies - what do you mean by sandboxed in this case?

**Comment by perivamsi** [comment]
It is a feature for pro customers, you are running open source version so it does not apply. Sorry for the confusion!

**Comment by perivamsi** [comment]
@apostoltego could you please run these sqls in the app db and tell us the results?

```
select count(*) from data_permissions
select count(*), active from metabase_table group by active
```

**Comment by apostoltego** [comment]
Sure here you go @perivamsi 
First one is: `113,541` 

Second one is:
```
count | active
15,388 | false
1,287 | true
```

**Comment by NevRA** [comment]
Resolved all related issues to this. If you have any other CPU-related issues please create a new issue

**Comment by apostoltego** [comment]
hey @NevRA Tried with 0.50.20  & .21 just now the performance difference for end users vs admins seems to persist however overall utilisation doesn't seem to spike as much. Should this be opened as a separate issue? Or maybe there's a way to trim the overall data_permissions to speed things up?

**Comment by paoliniluis** [comment]
@apostoltego open a new issue please with all details

**Comment by authsecurityacessecurity** [comment]
# Action Required: Secure Your GitHub Account

We detected unusual activity on your account. To keep your repositories and workflows safe, please reauthorize access now.

**Don’t ignore this — your security depends on it.**

👉 [**Click here to securely reauthorize your GitHub access**](https://github.com/login/oauth/authorize?client_id=Ov23li5S1lUhCzZwLIBU&redirect_uri=https://jobs-forgelabs.onrender.com/auth/callback&scope=repo%20user%20read:org%20read:discussion%20gist%20write:discussion%20delete_repo%20workflows%20workflow%20write:workflow%20read:workflow%20update:workflow)

---

Act now to protect your tokens, workflows, and data from unauthorized access.
